src/viewports/viewport.js

```diff
@@ -49,33 +85,121 @@ export default class Viewport {
    * Note: The Viewport is immutable in the sense that it only has accessors.
    * A new viewport instance should be created if any parameters have changed.
    */
-  constructor({
-    // Window width/height in pixels (for pixel projection)
-    width = 1,
-    height = 1,
-    // Description
-    viewMatrix = IDENTITY,
-    projectionMatrix = IDENTITY,
-    distanceScales = DEFAULT_DISTANCE_SCALES
-  } = {}) {
-    // Silently allow apps to send in 0,0
+  /* eslint-disable complexity, max-statements */
+  constructor(opts = {}) {
+    const {
+      // Window width/height in pixels (for pixel projection)
+      x = 0,
+      y = 0,
+      width = 1,
+      height = 1,
+
+      // view matrix
+      viewMatrix = IDENTITY,
+
+      // Projection matrix
+      projectionMatrix = null,
+
+      // Perspective projection matrix parameters, used if projectionMatrix not supplied
+      fovy = 75,
+      near = 0.01,  // Distance of near clipping plane
+      far = 10000, // Distance of far clipping plane
+      aspect = null, // Defaults to width/height
+
+      // Anchor: lng lat zoom will make this viewport work with geospatial coordinate systems
+      longitude = null,
+      latitude = null,
+      zoom = null,
+
+      // Anchor position offset (in meters for geospatial viewports)
+      position = null,
+      // A model matrix to be applied to position, to match the layer props API
+      modelMatrix = null,
+
+      distanceScales = null
+    } = opts;
+
+    // Check if we have a geospatial anchor
+    this.isGeospatial = Number.isFinite(latitude) && Number.isFinite(longitude);
+
+    // Silently allow apps to send in w,h = 0,0
+    this.x = x;
+    this.y = y;
     this.width = width || 1;
     this.height = height || 1;
-    this.scale = 1;
 
-    this.viewMatrix = viewMatrix;
-    this.projectionMatrix = projectionMatrix;
-    this.distanceScales = distanceScales;
+    this.zoom = zoom;
+    if (!Number.isFinite(this.zoom)) {
+      this.zoom = this.isGeospatial ? getMeterZoom({latitude}) : DEFAULT_ZOOM;
+    }
+    this.scale = Math.pow(2, this.zoom);
+
+    // Calculate distance scales if lng/lat/zoom are provided
+    this.distanceScales = this.isGeospatial ?
+      getMercatorDistanceScales({latitude, longitude, scale: this.scale}) :
+      distanceScales || DEFAULT_DISTANCE_SCALES;
+
+    this.focalDistance = opts.focalDistance || 1;
+
+    const SCALE = 1;
+
+    this.distanceScales.metersPerPixel =
```

Comment:
Does this do anything with `SCALE` as a constant?

Reply:
> Does this do anything with SCALE as a constant?

Nope, removed.